Expand Circuit Breaker Troubleshooting (#4520)
elastic/docs-content: troubleshoot/elasticsearch/circuit-breaker-errors.md

Caused by: org.elasticsearch.common.breaker.CircuitBreakingException: [parent] Data too large, data for [<transport_request>] would be [num/numGB], which is larger than the limit of [num/numGB], usages [request=0/0b, fielddata=num/numKB, in_flight_requests=num/numGB, accounting=num/numGB]
```

**Check circuit breaker statistics**

You can use the [node stats API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-stats) to get statistics about the circuit breaker per node.

```console
GET _nodes/stats?filter_path=nodes.*.breakers

> ```

Reviewer: @tanja-milicic an optional good add would be from [this request](https://github.com/elastic/elasticsearch/issues/132688) implemented [here](https://github.com/elastic/elasticsearch/pull/136890) starting in `9.3.0` users will have a CAT Circuit Breakers API which will make troubleshooting this much faster.

Author: Nice! That is definitely much cleaner than current approach.
How would we add that since it is not available yet? Would it be better to add the change once we have 9.3.0 released?

Reviewer 2: Okay, I see! I think in this situation, the easiest option is to use an [applies_to switch](https://docs-v3-preview.elastic.dev/elastic/docs-builder/tree/main/syntax/applies-switch).

The result would look like different tabs for your different versions:
The tab for 9.0 - 9.2:
<img width="935" height="365" alt="image" src="https://github.com/user-attachments/assets/849bd553-2a8d-45b5-b0a0-8ae5586b16aa" />

And the tab for 9.3:
<img width="935" height="365" alt="image" src="https://github.com/user-attachments/assets/8ec648cb-3aa8-43b9-9b01-cb081c5e33be" />

Here's the syntax I used:

-----
::::{applies-switch}

:::{applies-item} { "stack": "ga" }
Use the [get node statistics](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-stats) API:

```console
GET _nodes/stats?filter_path=nodes.*.breakers
```
:::

:::{applies-item} { "stack": "ga 9.3" }
Use the [get circuit breakers statistics](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-circuit-breaker) API:
```console
GET /_cat/circuit_breaker/
```
:::

::::

---------

<img width="1140" height="530" alt="image" src="https://github.com/user-attachments/assets/3eb603e7-2977-481f-91af-68390257412e" />

Reviewer 2: Maybe I can just push this commit to your branch if you don't mind? Might be a bit easier 😅

Reviewer 2: I pushed the tabs-related syntax, but now I'm not so sure the tabs work as they would imply the two API calls are mutually exclusive (which they aren't as far as I can tell, because the Node Stats API will still be useable, right? It's not being deprecated or anything?). 

Let me try just listing the two requests as options with in-line `applies_to` syntax.

Reviewer 2: I've done that in [1887f6e](https://github.com/elastic/docs-content/pull/4520/changes/1887f6ea45bf52283d95d11c1752be240e3761aa). I've listed both options and included descriptions for what each response will provide. Could you please review for accuracy?

```console
POST _cache/clear?fielddata=true
```

## Circuit Breaker Types

> [Circuit breaker types](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md) are manually defined for known expensive code paths with a sum-up [parent circuit breaker](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md#parent-circuit-breaker). Breakers need to be resolved per breaker type; common examples:

Author: Good point! I have removed it as most of it is in the Circuit breaker settings page. However, I see that we removed `Accounting requests circuit breaker` from the docs - it is available up until 8.13 version. I can't find the reason why it was removed. Any idea?

cc @yetanothertw

> [Circuit breaker types](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md) are manually defined for known expensive code paths with a sum-up [parent circuit breaker](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md#parent-circuit-breaker). Breakers need to be resolved per breaker type; common examples:

Reviewer 2: Hi @tanja-milicic, you can still find it for the 8.13 version of the docs here: https://www.elastic.co/guide/en/elasticsearch/reference/8.13/circuit-breaker.html#accounting-circuit-breaker

We're using a different versioning system for 9.x docs (which is the file you're editing now); it's a [cumulative docs approach](https://www.elastic.co/docs/contribute-docs/how-to/cumulative-docs).

Do you need to edit the 8.13 docs? I don't think we maintain the docs that far back, but I can ask my colleagues for ideas on how to proceed.

> [Circuit breaker types](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md) are manually defined for known expensive code paths with a sum-up [parent circuit breaker](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md#parent-circuit-breaker). Breakers need to be resolved per breaker type; common examples:

Reviewer 2: I think that would be [this PR here](https://github.com/elastic/elasticsearch/pull/107015). Note that it's located in the `elasticsearch` repo (as opposed to the `docs-content` repo that hosts 9.x content) on the 8.14 branch -- the 8.x content and branching strategy for docs is/was different.

> [Circuit breaker types](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md) are manually defined for known expensive code paths with a sum-up [parent circuit breaker](elasticsearch://reference/elasticsearch/configuration-reference/circuit-breaker-settings.md#parent-circuit-breaker). Breakers need to be resolved per breaker type; common examples:

Author: Ah right, thank you for explaining it!

- `fielddata`: [field data] and [global ordinals]
- `eql_sequence`: [eql sequences]
- `request`: API request bodies. Commonly for [Bulk] sizes too large
- `inflight_requests`: total API network bytes
- `script`: [scripts]
- `regex`: [regex]

**Memory evaluation**

Circuit breakers may either directly evaluate memory usage estimates or indirectly limit operations that are likely to cause excessive memory consumption. For example, the `script` circuit breaker checks memory indirectly by rate-limiting Painless/Mustache script compilations. However, even with circuit breakers in place, nodes can still encounter out-of-memory (OOM) conditions. This can occur, for example, because:

- Circuit breaker relies on point-in-time memory usage estimations.
- Parallel operations may still heap DOS-attack the node even with `parent` circuit breakers.
- Certain dynamic operations can quickly consume substantial memory. For example [aggregations](elasticsearch://explore-analyze/query-filter/aggregations.md) and [complex queries](elasticsearch://reference/query-languages/query-dsl/compound-queries).

> Circuit breakers protect JVM heap memory, but if OOM occurs due to non-heap memory; for example JVM for complication (code cache) or thread stacks.

Reviewer: This sentence should finish not hang (both in KB as well as here). My guess on intended message is
```suggestion
Circuit breakers protect the node's JVM heap. OOM can still trigger due to non-heap memory, for example within the compilation or thread stacks.
```
